plugin/core/url.py

```diff
@@ -1,19 +1,63 @@
+from .typing import Any
+from urllib.parse import quote
 from urllib.parse import urljoin
 from urllib.parse import urlparse
 from urllib.request import pathname2url
 from urllib.request import url2pathname
 import os
+import re
 
+import sublime
 
-def filename_to_uri(path: str) -> str:
-    return urljoin('file:', pathname2url(path))
+
+def filename_to_uri(file_name: str) -> str:
+    """
+    Convert a file name obtained from view.file_name() into an URI
+    """
+    prefix = sublime.installed_packages_path()
+    if file_name.startswith(prefix):
+        return _to_resource_uri(file_name, prefix)
+    prefix = sublime.packages_path()
+    if file_name.startswith(prefix) and not os.path.exists(file_name):
+        return _to_resource_uri(file_name, prefix)
+    path = pathname2url(file_name)
+    re.sub(r"^([A-Z]):/", _lowercase_driveletter, path)
+    return urljoin("file:", path)
+
+
+def view_to_uri(view: sublime.View) -> str:
+    file_name = view.file_name()
+    if not file_name:
+        return "buffer://sublime/{}".format(view.buffer_id())
+    return filename_to_uri(file_name)
 
 
 def uri_to_filename(uri: str) -> str:
+    """
+    DEPRECATED: An URI associated to a view does not necessarily have a "file:" scheme.
+    Use urllib.parse.urlparse to determine the scheme and go from there.
+    Use urllib.parse.unquote to unquote the path.
+    """
     parsed = urlparse(uri)
     assert parsed.scheme == "file"
     if os.name == 'nt':
         # url2pathname does not understand %3A (VS Code's encoding forced on all servers :/)
         return url2pathname(parsed.path).strip('\\')
     else:
         return url2pathname(parsed.path)
+
+
```

Comment:
Can we have an alternative helper function for getting filename from URI? I'd need to use something in eslint.

Having to play with urllib doesn't sound like fun and also I wonder if such helper needs to apply some path transformations to stay consistent.

I imagine such helper could return a tuple with scheme and file path (if `file` scheme at least).

Reply:
Let's also do that in another PR?